src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs

```diff
@@ -880,6 +880,30 @@ static void StoreNullableIntArray(HttpContext httpContext, int?[]? tryParsable)
         Assert.Null(httpContext.Items["tryParsable"]);
     }
 
+    [Fact]
+    public async Task RequestDelegateHandlesOptionalStringValuesFromNullQueryString()
+    {
+        var httpContext = CreateHttpContext();
+        httpContext.Request.Query = new QueryCollection(new Dictionary<string, StringValues>
+        {
+            ["tryParsable"] = (string?)null
+        });
+
+        static void StoreNullableIntArray(HttpContext httpContext, StringValues? tryParsable)
```

Comment:
On balance I think that `StringValues` for _required_ and `StringValues?` for _optional_ makes the most sense overall as that's more in-keeping with the behaviour for `int` vs. `int?` etc.

`StringValues.Empty` is just an anomaly, as it were, as most other structs don't have such a thing.

Then, like you say, the endpoint code would be something like this for someone's implementation for required vs. optional:

```csharp
app.MapGet("/foo", (StringValues myRequiredHeader) =>
{
    return Results.Json(new { count = myRequiredHeader.Count });
});

app.MapGet("/bar", (StringValues? myOptionalHeader) =>
{
    return Results.Json(new { count = myOptionalHeader.HasValue ? myOptionalHeader.Value.Count : -1 });
});
```

Reply:
The only thing I don't like about supporting nullable `StringValues` is it ends up meaning that we *never* pass in `StringValues.Empty`. It's always `null` or a non-empty `StringValues`.

If all I was concerned about was the number of a given headers (0, 1, 2, whatever), this would be annoying. I imagine that use case is super rare though, so I still think we should treat non-nullable `StringValues` as required.